Removed lib/system/allocators.nim. seqs_v2 and strs_v2 now uses allocShared0.

zevv opened the pull request:
Creating a draft PR for CI. Comments are welcome though.

zevv committed:
Cleanup, remove lib/system/allocators.nim. seqs_v2 and strs_v2 now use
allocShared0 by default.

Araq commented on lib/system/alloc.nim:
```diff
@@ -997,13 +997,21 @@ template instantiateForRegion(allocator: untyped) {.dirty.} =
 
   proc deallocOsPages = deallocOsPages(allocator)
 
+  proc c_getenv(env: cstring): cstring {.importc: "getenv", header: "<stdlib.h>".}
```
Ha ha, no...

zevv replied:
Yeah, gone now, sorry for that.

zevv committed:
Fixed -d:useMalloc allocShared / reallocShared / deallocShared. These now use the alloc/dealloc/realloc implementation that also takes care of zeroing memory at realloc.

zevv committed:
Removed debug printfs

zevv commented:
I also fixed a mixup in the `useMalloc` versions of  `[de|re]allocShared()` and friends, some of those used the lenght-prefixed version and some didn't.

zevv committed:
Removed unpairedEnvAllocs() from tests/destructor/tnewruntime_misc

zevv committed:
More mmdisp cleanups. The shared allocators do not need to zero memory or throw since the regular ones already do that

zevv commented:
I'm not sure why, but I have a failing test on 32 bits with --gc:arc getting a SIGSEGV. Interestingly, this happens in a test which was added because of a SIGSEGV: #12964 which was fixed by  #13027.

Reproduce with

```
nim_temp c --passC:-m32 --passL:-m32 --cpu:i386 --gc:arc tests/destructor/tarc.nim
```

This is the trace. It is not exactly the same as #12964, but the final crash also happens in `nimDestroyAndDispose`. I do not understand the fix from #13027 though, and I'm not sure if/how his is related.

Trace:

```
@[1, 2, 3]
Success
@["a", "b", "c"]
==3604850== Invalid read of size 4
==3604850==    at 0x111787: nimDestroyAndDispose (in /home/ico/external/Nim/tests/destructor/tarc)
==3604850==    by 0x117820: eqsink___yQpfZnOhtVx1qY5Xexf1UQ (in /home/ico/external/Nim/tests/destructor/tarc)
==3604850==    by 0x112D09: add__q2FXfIRF6XuhcfFeO9cmSTQ (in /home/ico/external/Nim/tests/destructor/tarc)
==3604850==    by 0x1179E7: bug12964__68KbLG1sEj1akMRruR9bUqQ_8 (in /home/ico/external/Nim/tests/destructor/tarc)
==3604850==    by 0x11827F: NimMainModule (in /home/ico/external/Nim/tests/destructor/tarc)
==3604850==    by 0x1180D1: NimMainInner (in /home/ico/external/Nim/tests/destructor/tarc)
==3604850==    by 0x118108: NimMain (in /home/ico/external/Nim/tests/destructor/tarc)
==3604850==    by 0x118146: main (in /home/ico/external/Nim/tests/destructor/tarc)
==3604850==  Address 0x1 is not stack'd, malloc'd or (recently) free'd
==3604850== 
Traceback (most recent call last)
tarc.nim(118)            tarc
tarc.nim(116)            bug12964
refs_v2.nim(116)         nimDestroyAndDispose
SIGSEGV: Illegal storage access. (Attempt to read from nil?)
```

disruptek commented:
My only idea is that the offset for Cell is off because the header size is different, but I don't see how we'd be getting `0x1`.  Plus, don't we increment with `0b100` or whatever?

Anyway, now I'm running your branch and I cannot reproduce any results.
When I run the test, I get the following:
```
@[1, 2, 3]
Success
@["a", "b", "c"]
Traceback (most recent call last)
tarc.nim(118)            tarc
tarc.nim(116)            bug12964
refs_v2.nim(133)         nimDecRefIsLast
SIGSEGV: Illegal storage access. (Attempt to read from nil?)
```
When I tried to run it with valgrind, I got a SIGILL
```
vex x86->IR: unhandled instruction bytes: 0xC4 0xE2 0x7A 0xF7
==22899== valgrind: Unrecognised instruction at address 0x400ba43.
==22899==    at 0x400BA43: _dl_new_object (in /lib32/ld-2.29.so)
==22899==    by 0x4002BFA: dl_main (in /lib32/ld-2.29.so)
==22899==    by 0x4018DE5: _dl_sysdep_start (in /lib32/ld-2.29.so)
==22899==    by 0x4001F5A: _dl_start (in /lib32/ld-2.29.so)
==22899==    by 0x40010BA: ??? (in /lib32/ld-2.29.so)
```

This is my `nim.cfg`:
```
--debuginfo
--passC="-m32"
--passL="-m32"
--cpu:i386
hint[Processing]=off
hint[Conf]=off
-d:release
-d:danger
-d:exceptions:goto
--embedsrc
```

Araq commented on lib/system/strs_v2.nim:
```diff
@@ -41,18 +43,15 @@ proc prepareAdd(s: var NimStringV2; addlen: int) {.compilerRtl.} =
-    s.p = cast[ptr NimStrPayload](s.p.allocator.realloc(s.p.allocator, s.p,
-      oldSize = contentSize(s.p.cap),
-      newSize = contentSize(cap)))
+    s.p = cast[ptr NimStrPayload](reallocShared(s.p, contentSize(cap)))
```
The old variant was much better as it remembered the old size and zeroed out the additional memory.

zevv replied:
Fixed by introducing `realloc[Shared]0()`

zevv committed:
Introduced realloc0 and reallocShared0, these procs are now used by
strs_v2 and seqs_v2. This also allowed the -d:useMalloc allocator to
drop the extra header with allocation length.

Araq commented on lib/system/mmdisp.nim:
```diff
@@ -274,6 +275,9 @@ elif defined(gogc):
     doAssert false, "not implemented"
 
+  proc realloc0(p: pointer, oldsize, newsize: Natural): pointer =
+    doAssert false, "not implemented"
```
It's better to not implement this via ``.error`` rather than fail at runtime.

zevv replied:
Done

Araq commented:
Instead of this ``allocated`` field, I'd prefer if we could do without and encode it via an unused bit in "capacity" like the old runtime does, see ``NIM_STRLIT_FLAG``. Would save 8 bytes for every string and sequence.

zevv committed:
goGc now uses .error to report unimplemented calls

zevv committed:
Moved strs_v2/seqs_v2 'allocated' flag into 'cap' field

zevv reviewed (commented):
`allocated` field dropped

zevv committed:
Revert "goGc now uses .error to report unimplemented calls". This breaks
CI

This reverts commit 53c89ee5603730fa9f4bab0b17d34266acab9605.

Araq commented on tests/destructor/tnewruntime_strutils.nim:
```diff
@@ -214,5 +212,3 @@ staticTests()
 let xaa = @[""].join()
 let xbb = @["", ""].join()
 
-let (a, d) = allocCounters()
```
This needs to check for ``getOccupiedMem() == 0`` instead then so that we have some leak detection for the targets that happen to not use ``valgrind``. Hmm but then the ``-d:useMalloc`` does bite. So we need a testament patch, ``valgrind: true`` should imply ``-d:useMalloc``.

Araq commented on tests/destructor/tv2_raise.nim:
```diff
@@ -50,6 +48,3 @@ except:
 
 echo "OK ", ok
-
-let (a, d) = allocCounters()
```
Same here, needs some leak detection.

Araq commented on tests/destructor/twidgets.nim:
```diff
@@ -74,5 +72,3 @@ proc main =
 
 main()
 
-let (a, d) = allocCounters()
```
Same here.

Araq commented on tests/destructor/twidgets_unown.nim:
```diff
@@ -71,5 +69,3 @@ proc main =
 
 main()
 
-let (a, d) = allocCounters()
```
Same here.

Commit:
Added 'getAllocStats()' to get low level alloc/dealloc counters. Enable with -d:allocStats

zevv commented:
Proposal for allocation statistics in order to check for leaks in tests: 

https://github.com/nim-lang/Nim/commit/20bb05b6cd7db6af367b54fb97050fca45d7bc93

Commit:
*allocShared implementations for boehm and go allocators now depend on the proper *allocImpl procs

zevv commented:
Alloc/dealloc counters are in place, enable with `-d:allocStats`. The reported numbers are a bit messy because all Nim allocations are now counted, not only seqs/strings.

Araq commented on tests/destructor/tgcdestructors.nim:
```diff
@@ -10,10 +10,9 @@ a: @[4, 2, 3]
 30
 true
-41 41'''
+(allocCount: 123, deallocCount: 82)'''
```
So we leak 41 objects? How come?

Clyybber replied:
I think we alloc more often than dealloc due to string/seq resizing

zevv replied:
Yeah, I mentioned this on #nim already, and I'm not sure what is the best way to solve this.

This implementation keeps track of all allocs, not only allocs made by seq_v2/str_v2. There is still memory alive when the statistics are gathered and printed. I guess the only way to solve this is to wait until after Nim is all done, somewhere after NimMainInit()

Araq commented on tests/destructor/tnewruntime_misc.nim:
```diff
@@ -1,17 +1,15 @@
 destroying GenericObj[T] GenericObj[system.int]
 test
-'''
+(allocCount: 34, deallocCount: 15)'''
```
Again, why this difference?

zevv commented:
Ok, there is some double accounting done still, fixing...

zevv commented:
Fixed the counters, the allocShared implementations were recursively calling back into the non-shared versions. Sorry for the noise.

There are two asymmetric cases left now:

- tests/destructor/tv2_raise.nim
- tests/destructor/tnewruntime_misc.nim

dos not leak, but have allocations that get destroyed only at the end of `NimMainModule`

Araq commented:
Optional follow-up improvements can be done later, merging.
